src/main/java/com/onarandombox/MultiverseCore/listeners/MVPlayerListener.java

```diff
@@ -144,87 +167,54 @@ public void playerChangedWorld(PlayerChangedWorldEvent event) {
+
+        if (teleporter == null) {
+            Logging.finer("We don't know who the teleporter is, so we'll assume it's the player.");
+            teleporter = teleportee;
         }
 
-        // Check if player is allowed to enter the world if we're enforcing permissions
-        if (plugin.getMVConfig().getEnforceAccess()) {
-            event.setCancelled(!pt.playerCanGoFromTo(fromWorld, toWorld, teleporter, teleportee));
-            if (event.isCancelled() && teleporter != null) {
-                Logging.fine("Player '" + teleportee.getName()
-                        + "' was DENIED ACCESS to '" + toWorld.getAlias()
-                        + "' because '" + teleporter.getName()
-                        + "' don't have: multiverse.access." + event.getTo().getWorld().getName());
-                return;
-            }
-        } else {
-            Logging.fine("Player '" + teleportee.getName()
-                    + "' was allowed to go to '" + toWorld.getAlias() + "' because enforceaccess is off.");
-        }
-
-        // Does a limit actually exist?
-        if (toWorld.getPlayerLimit() > -1) {
-            // Are there equal or more people on the world than the limit?
-            if (toWorld.getCBWorld().getPlayers().size() >= toWorld.getPlayerLimit()) {
-                // Ouch the world is full, lets see if the player can bypass that limitation
-                if (!pt.playerCanBypassPlayerLimit(toWorld, teleporter, teleportee)) {
-                    Logging.fine("Player '" + teleportee.getName()
-                            + "' was DENIED ACCESS to '" + toWorld.getAlias()
-                            + "' because the world is full and '" + teleporter.getName()
-                            + "' doesn't have: mv.bypass.playerlimit." + event.getTo().getWorld().getName());
-                    event.setCancelled(true);
-                    return;
-                }
-            }
+        MVWorld fromWorld = this.worldManager.getMVWorld(event.getFrom().getWorld());
+        MVWorld toWorld = this.worldManager.getMVWorld(event.getTo().getWorld());
+        ActionResponse actionCheckResponse = this.actionChecker.canGoFromToWorld(teleporter, teleportee, fromWorld, toWorld);
+        Logging.fine(actionCheckResponse.toString());
+        if (!actionCheckResponse.isSuccessful()) {
+            tellReason(teleporter, teleportee, fromWorld, toWorld, actionCheckResponse);
+            event.setCancelled(true);
+            return;
         }
 
-        // By this point anything cancelling the event has returned on the method, meaning the teleport is a success \o/
-        this.stateSuccess(teleportee.getName(), toWorld.getAlias());
-    }
+        if (actionCheckResponse.hasResult(EntryFeeResult.ENOUGH_MONEY)) {
+            double price = toWorld.getPrice();
+            Material currency = toWorld.getCurrency();
+            this.plugin.getEconomist().payEntryFee((Player) teleporter, price, currency);
+        }
```

Comment:
Kind of wondering if we should move this to a separate MONITOR version of this event so we know that the player has actually been teleported.

Reply:
We could do that yea, but i think not the scope of this PR. Prob do it when i redo the teleport queue stuff, which is really bad. Its like static methods in MultiverseCore class

Reply:
Want to track this plan somewhere and then we can resolve this convo?

Reply:
ig setting up github project maybe?

Reply:
https://github.com/orgs/Multiverse/projects/4/views/1?pane=issue&itemId=21658559

Reply:
nice, will add to the project page